File: gachisa-backend/src/test/java/com/gachisa/auth/service/RefreshTokenServiceTest.java
package com.gachisa.auth.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

import com.gachisa.auth.repository.RefreshTokenCacheRepository;
import com.gachisa.auth.repository.RefreshTokenRepository;
import com.gachisa.global.security.JwtProperties;
import com.gachisa.global.security.TokenHashUtil;
import com.gachisa.user.service.UserService;
import java.time.LocalDateTime;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class RefreshTokenServiceTest {

    private static final LocalDateTime CUTOFF = LocalDateTime.of(2026, 7, 25, 4, 0);

    @Mock
    private RefreshTokenRepository refreshTokenRepository;

    @Mock
    private RefreshTokenCacheRepository refreshTokenCacheRepository;

    @Mock
    private UserService userService;

    @Mock
    private TokenHashUtil tokenHashUtil;

    @Mock
    private JwtProperties jwtProperties;

    private RefreshTokenService refreshTokenService;

    @BeforeEach
    void setUp() {
        refreshTokenService = new RefreshTokenService(
                refreshTokenRepository, refreshTokenCacheRepository, userService, tokenHashUtil, jwtProperties);
    }

    @Test
    void cleanupStaleTokensDelegatesToRepositoryAndReturnsDeletedCount() {
        given(refreshTokenRepository.deleteAllExpiredBefore(CUTOFF)).willReturn(5);

        int deletedCount = refreshTokenService.cleanupStaleTokens(CUTOFF);

        assertThat(deletedCount).isEqualTo(5);
        verify(refreshTokenRepository).deleteAllExpiredBefore(CUTOFF);
    }
}

File: gachisa-backend/src/test/kotlin/com/gachisa/auth/service/RefreshTokenServiceTest.kt
package com.gachisa.auth.service

import com.gachisa.auth.repository.RefreshTokenCacheRepository
import com.gachisa.auth.repository.RefreshTokenRepository
import com.gachisa.global.security.JwtProperties
import com.gachisa.global.security.TokenHashUtil
import com.gachisa.user.service.UserService
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class RefreshTokenServiceTest {

    private val refreshTokenRepository: RefreshTokenRepository = mockk()
    private val refreshTokenCacheRepository: RefreshTokenCacheRepository = mockk()
    private val userService: UserService = mockk()
    private val tokenHashUtil: TokenHashUtil = mockk()
    private val jwtProperties: JwtProperties = mockk()

    private lateinit var refreshTokenService: RefreshTokenService

    @BeforeEach
    fun setUp() {
        refreshTokenService = RefreshTokenService(
            refreshTokenRepository, refreshTokenCacheRepository, userService, tokenHashUtil, jwtProperties,
        )
    }

    @Test
    fun cleanupStaleTokensDelegatesToRepositoryAndReturnsDeletedCount() {
        every { refreshTokenRepository.deleteAllExpiredBefore(CUTOFF) } returns 5

        val deletedCount = refreshTokenService.cleanupStaleTokens(CUTOFF)

        assertThat(deletedCount).isEqualTo(5)
        verify { refreshTokenRepository.deleteAllExpiredBefore(CUTOFF) }
    }

    companion object {
        private val CUTOFF: LocalDateTime = LocalDateTime.of(2026, 7, 25, 4, 0)
    }
}
